### PR TITLE
KATI-17_Implement-Register-Patient-Function

### PR DESCRIPTION
**Pull request by vtnlinh95**

### CODE REVIEW

File: app/src/main/java/com/kms/cura/view/PatientSignUpActivity.java [hunk 1]
```diff
@@ -28,7 +27,8 @@ protected void onCreate(Bundle savedInstanceState) {
         btnRegister.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
-                //Patient Register Function
+                UserEntity entity = new UserEntity(null, getEditTextText(edtFirstName), getEditTextText(edtEmail), getEditTextText(edtPassword));
+                UserModel.getInstace().registerPatient(entity,PatientSignUpActivity.this);
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/PatientSignUpActivity.java]: theses 2 lines belongs to UserController

File: com.kms.cura_server/src/com/kms/cura_server/UserAPI.java [hunk 2]
```diff
@@ -105,4 +110,16 @@ private Type getPatientUserType() {
 		}.getType();
 		return type;
 	}
+	private String SuccessResponse(UserEntity user){
+		JsonElement JsonUser = EntityToJsonConverter.convertEntityToJson(user);
+		JsonObject JsonUserConvert =(JsonObject) JsonUser;
+		JsonUserConvert.addProperty(STATUS_KEY, statusKey);
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: should set STATUS_KEY here to 'true' or to a constant.

File: com.kms.cura_server/src/com/kms/cura_server/UserAPI.java [hunk 3]
```diff
@@ -105,4 +110,16 @@ private Type getPatientUserType() {
...
+		return JsonUserConvert.toString();
+	}
+	private String UnsuccessResponse(Exception e){
+		JsonObject jsonError=new JsonObject();
+		jsonError.addProperty(STATUS_KEY, !statusKey);
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: should set STATUS_KEY here to 'true' or to a constant.

File: com.kms.cura_server/src/com/kms/cura_server/UserAPI.java [hunk 4]
```diff
@@ -19,10 +20,14 @@
 import com.kms.cura.entity.json.JsonToEntityConverter;
 import com.kms.cura.entity.user.DoctorUserEntity;
 import com.kms.cura.entity.user.PatientUserEntity;
+import com.kms.cura.entity.user.UserEntity;
 import com.kms.cura_server.resources.Strings;
 
 @Path("/user")
 public final class UserAPI {
+	private boolean statusKey = true;
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: this should be removed for usage of constant

File: com.kms.cura_server/src/com/kms/cura_server/UserAPI.java [hunk 5]
```diff
@@ -19,10 +20,14 @@
...
+	private String STATUS_KEY = "statuskey";
+	private String MESSAGE= "message";
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: MESSAGE and STATUS_KEY should be move down to com.kms.entity for re-usable between client and server

File: app/src/main/java/com/kms/cura/model/UserModel.java [hunk 6]
```diff
@@ -27,4 +41,41 @@ public boolean save(UserEntity user) {
...
+                    JSONObject jsonObject = new JSONObject(response);
+                    boolean status = jsonObject.getBoolean(STATUS_KEY);
+                    if (status) {
+                        ErrorController.showDialog(mContext,"success");
+                    } else {
+                        String message = jsonObject.getString(MESSAGE);
+                        ErrorController.showDialog(mContext, message);
+                    }
+                } catch (JSONException e) {
+                    e.printStackTrace();
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/model/UserModel.java]: Remove printStackTrace here

File: app/src/main/java/com/kms/cura/model/Settings.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.kms.cura.model;
+
+/**
+ * Created by linhtnvo on 6/6/2016.
+ */
+public class Settings {
+    public static final String URL_CREATE_PATIENT_API = "http://10.0.3.2:8080/com.kms.cura_server-0.1/cura_server/user/createPatient";
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/model/Settings.java]: should separate URL_CREATE_PATIENT_API into SERVER_URL and CREATE_PATIENT_API for re-use

File: app/src/main/java/com/kms/cura/model/UserModel.java [hunk 8]
```diff
@@ -27,4 +41,41 @@ public boolean save(UserEntity user) {
         // call to api to save new user
         return false;
     }
+    public void registerPatient(final UserEntity Entity, final Context mContext){
+        StringRequest stringRequest = new StringRequest(Request.Method.POST, Settings.URL_CREATE_PATIENT_API, new Response.Listener<String>() {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/model/UserModel.java]: Should extract new Response.Listener<String>() to a new class for later re-use with other entity types

Comment (hieumai) [inline app/src/main/java/com/kms/cura/model/UserModel.java]: Should also extract new StringRequest to a class for re-use

File: com.kms.cura_server/src/com/kms/cura_server/UserAPI.java (hunk 4 above)
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: Also, we only have 2 status: success or fail ?

Comment (vtnlinh95) [inline com.kms.cura_server/src/com/kms/cura_server/UserAPI.java]: Yes, we have many kinds of error but all of them are considered as unsuccessful creating patient

File: app/src/main/java/com/kms/cura/controller/UserController.java [hunk 9]
```diff
@@ -1,9 +1,15 @@
 package com.kms.cura.controller;
 
+import android.content.Context;
+
 import com.kms.cura.entity.user.UserEntity;
 import com.kms.cura.model.UserModel;
 
 public class UserController {
+    public static void registerPatient(Context mContext, String name, String email, String password){
+        UserEntity entity=new UserEntity(null, name, email,password);
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/controller/UserController.java]: Should re-format this file

File: app/src/main/java/com/kms/cura/model/UserModel.java [hunk 10]
```diff
@@ -34,48 +33,22 @@ public static UserModel getInstace() {
...
-                JsonElement jsonElement = EntityToJsonConverter.convertEntityToJson(Entity);
-                return jsonElement.toString().getBytes();
-            }
 
-            @Override
-            public String getBodyContentType() {
-                return RAW_TYPE;
-            }
-        };
+    public void registerPatient(final UserEntity Entity, final Context mContext) {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/model/UserModel.java]: variable "Entity" should be in lowercase

**Comment by vtnlinh95** [comment]
Already fixed them

**Comment by hieumai** [comment]
Ok. Linh please rebase this into 1 commit. Meanwhile, @toannguyen71195  and @huuduy216 please start verifying this ticket.

**Comment by huuduy216** [comment]
@vtnlinh95  
- The first field in patient register should be "Name" not "First Name" (since we only have name in the database). The validation for that field will return invalid as soon as there's a space in the name ( which is undesired, in my opinion).
- Server address should to be changed to actually server address instead of you machine (192.168.74.141:8080) for better consistency in client as well as server/api testing
-  The "success" message shoud be "Success" for better looking.
-  Also all of the errors should have a messages associate with them instead just the error name in java as of right now.
  
  Other than those things great work :D

**Comment by vtnlinh95** [comment]
@huuduy216 I follow the mockup which is "First Name" so if you want to change please check with Jenny.

**Comment by huuduy216** [comment]
@vtnlinh95 : I've just discussed that with Jenny, we agreed that we'll only support name in our database, so it should be name only.

**Comment by toannguyen71195** [comment]
Some bugs found but have been fixed in person so VERIFIED.

**Comment by vtnlinh95** [comment]
The "success" message is used just for testing, when we have done patient profile view, it will directly go to that page instead of showing the "success" message

**Pull request merged**